### DISCUSSION TITLE
Mixed models: How to generate predictions "fixed" (marginalised?) over the random factors

### DISCUSSION THREAD

**Discussion by DominiqueMakowski** [TuringLang/Turing.jl#2390] [Q&A]
Let's start with a simple linear mixed model that takes `participant` as random factors.

```jl
using Turing, StatsFuns, DataFrames, Random

# Generate correlated x and y data
data = DataFrame(x=1:100, y=1:100, participant=repeat(1:10, inner=10))
data.y = data.x .+ randn(100) .* 10

# Linear
@model function model_Linear(y, x, participant, n)

    # Priors - Fixed Effects
    μ_intercept ~ Normal(0, 0.5)
    μ_x ~ Normal(0, 0.5)

    σ ~ truncated(Normal(0.0, 1), lower=0)

    # Priors - Random Effects
    μ_intercept_random_sd ~ truncated(Normal(0.0, 0.3), lower=0)
    μ_x_random_sd ~ truncated(Normal(0, 0.3), lower=0)

    # Participant-level priors
    μ_intercept_random ~ filldist(Normal(0, μ_intercept_random_sd), n)
    μ_x_random ~ filldist(Normal(0, μ_x_random_sd), n)

    for i in 1:length(y)
        # Compute μ
        μ = μ_intercept + μ_intercept_random[participant[i]]
        μ += (μ_x + μ_x_random[participant[i]]) * x[i, 1]

        # Likelihood
        y[i] ~ Normal(μ, σ)
    end
end

posteriors = sample(model_Linear(data.y, data.x, data.participant, 10), NUTS(), 200)
```

I can easily generate predictions on the original data, useful for instance for posterior predictive checks. 

```jl
# Predictions on whole dataset
pred = predict(model_Linear(fill(missing, 100), data.x, data.participant, 10), posteriors)
```

However, often for visualization purposes / postprocessing (marginal effects computation) we would need to generate predictions on new data, for instance on a data "grid" with specific values. While these values are often specifically chosen for the fixed effects, we are often interested in "marginalising" over the random effects, i.e., having the average prediction for the average random factor. 

I thought that one way to go about it would be to also set `participant` as `missing`, but that doesn't seem to be the correct way.

```jl
# Predictions on new data "grid"
grid = DataFrame(x=[0, 50, 100], participant=[missing, missing, missing])

pred2 = predict(model_Linear(fill(missing, 3), grid.x, grid.participant, 3), posteriors)
```

Any advice on how to achieve something like that?

**Comment by torfjelde** [comment]
> we are often interested in "marginalising" over the random effects, i.e., having the average prediction for the average random factor.

Maybe I'm misunderstanding, but do you just want

```julia
μ_intercept + μ_x * x[i, 1]
```

?

If you want to marginalize over the participants, then you need a notion of what the distribution of the participants are (which it doesn't seem like you have here, IIUC)

**Reply by DominiqueMakowski** [reply]
Ideally, I want `μ` assuming that `μ_x_random[participant[i]]=0` and `μ_intercept_random[participant[i]] = 0` (basically, predictions without the deviations from the fixed effects due to specific participants)

**Reply by DominiqueMakowski** [reply]
In other words, I'd like to nullify, for the predictions, any of the influence created by `participant` if that makes sense

**Reply by DominiqueMakowski** [reply]
Basically, the equivalent of *brms*'  `predict(brms_model, re_formula=~0)` (https://www.rdocumentation.org/packages/brms/versions/2.22.0/topics/predict.brmsfit)

**Reply by torfjelde** [reply]
> Ideally, I want μ assuming that μ_x_random[participant[i]]=0 and μ_intercept_random[participant[i]] = 0 (basically, predictions without the deviations from the fixed effects due to specific participants)

But this is equivalent to just
```julia
μ_intercept + μ_x * x[i, 1]
```
no? 

If you write out the expression but nullify the participant-specific params, then that's what you're left with.

**Reply by torfjelde** [reply]
If you just want to extract this, then I recommend doing the following:
```julia
@model function model_Linear(y, x, participant, n)

    # Priors - Fixed Effects
    μ_intercept ~ Normal(0, 0.5)
    μ_x ~ Normal(0, 0.5)

    σ ~ truncated(Normal(0.0, 1), lower=0)

    # Priors - Random Effects
    μ_intercept_random_sd ~ truncated(Normal(0.0, 0.3), lower=0)
    μ_x_random_sd ~ truncated(Normal(0, 0.3), lower=0)

    # Participant-level priors
    μ_intercept_random ~ filldist(Normal(0, μ_intercept_random_sd), n)
    μ_x_random ~ filldist(Normal(0, μ_x_random_sd), n)

    # CHANGED: Vectorized computation of the means.
    μ_fixed = μ_intercept .+ μ_x .* x[:, 1]
    μ_random = μ_intercept_random[participant] .+ μ_x_random[participant] .* x[:, 1]
    μ = μ_fixed .+ μ_random

    # Likelihood
    y .~ Normal.(μ, σ)

    # NOTE: Return the values so we can use `generated_quantities` to extract it.
    return (; μ_fixed, μ_random)
end


# Then once we have a chain, we can do the following to extract the values.
results = generated_quantities(model, chain)
```

**Reply by torfjelde** [reply]
You can also just insert the following if you want to make the random effects computation conditional on participant information being available, e.g.
```julia
μ = μ_fixed
if participant !== nothing
    μ = μ + μ_intercept_random[participant] .+ μ_x_random[participant] .* x[:, 1]
end
```
and remove references to `μ_random`

**Reply by DominiqueMakowski** [reply]
Thanks Tor as always

## 1. Vectorization issue

Unfortunately, it seems like when the model is in the vectorized form predicitons don't work (?)

```jl
@model function model_Linear(y, x, participant, n)

    # Priors - Fixed Effects
    μ_intercept ~ Normal(0, 0.5)
    μ_x ~ Normal(0, 0.5)

    σ ~ truncated(Normal(0.0, 1), lower=0)

    # Priors - Random Effects
    μ_intercept_random_sd ~ truncated(Normal(0.0, 0.3), lower=0)
    μ_x_random_sd ~ truncated(Normal(0, 0.3), lower=0)

    # Participant-level priors
    μ_intercept_random ~ filldist(Normal(0, μ_intercept_random_sd), n)
    μ_x_random ~ filldist(Normal(0, μ_x_random_sd), n)

    # Equation
    μ_fixed = μ_intercept .+ μ_x .* x[:, 1]
    μ_random = μ_intercept_random[participant] .+ μ_x_random[participant] .* x[:, 1]
    μ = μ_fixed .+ μ_random

    # Likelihood
    y .~ Normal.(μ, σ)
end

# Generate correlated x and y data
data = DataFrame(x=1:100, y=1:100, participant=repeat(1:10, inner=10))
data.y = data.x .+ randn(100) .* 10

# Sample
post = sample(model_Linear(data.y, data.x, data.participant, 10), NUTS(), 200)

# Predictions on whole dataset
pred = predict(model_Linear(fill(missing, 100), data.x, data.participant, 10), post)
```
```
ERROR: MethodError: no method matching loglikelihood(::Normal{Float64}, ::Missing)
The function `loglikelihood` exists, but no method is defined for this combination of argument types.
```


## 2. Using generated_quantities

You are correct I think, `μ = μ_intercept + μ_intercept_random[participant[i]] + (μ_x + μ_x_random[participant[i]]) * x[i, 1]` can be decomposed as:

```jl
μ_fixed = μ_intercept + μ_x * x[i, 1]
μ_random = μ_intercept_random[participant[i]] + μ_x_random[participant[i]] * x[i, 1]
μ = μ_fixed + μ_random
```

and thus `μ_fixed` is what I'm after. However, I am not sure how to access it *during prediction on a new dataset* (a new data grid of specific values of `x`). As far as I understand, `generated_quantities()` allows me to access the quantities generated during sampling? 


## 3. Conditional inclusion if random part of equation

Finally, following your suggestion, I wrote the following model:


```jl
@model function model_Linear(y, x, participant, n)

    # Priors - Fixed Effects
    μ_intercept ~ Normal(0, 0.5)
    μ_x ~ Normal(0, 0.5)

    σ ~ truncated(Normal(0.0, 1), lower=0)

    # Priors - Random Effects
    μ_intercept_random_sd ~ truncated(Normal(0.0, 0.3), lower=0)
    μ_x_random_sd ~ truncated(Normal(0, 0.3), lower=0)

    # Participant-level priors
    μ_intercept_random ~ filldist(Normal(0, μ_intercept_random_sd), n)
    μ_x_random ~ filldist(Normal(0, μ_x_random_sd), n)

    for i in 1:length(y)
        # Compute μ
        μ = μ_intercept + μ_x * x[i, 1]
        if !ismissing(participant[i])
            μ += μ_intercept_random[participant[i]] + μ_x_random[participant[i]] * x[i, 1]
        end

        # Likelihood
        y[i] ~ Normal(μ, σ)
    end
end
```

While it  works for original-data predictions:

```jl
# Generate correlated x and y data
data = DataFrame(x=1:100, y=1:100, participant=repeat(1:10, inner=10))
data.y = data.x .+ randn(100) .* 10

# Sample and predict
post = sample(model_Linear(data.y, data.x, data.participant, 10), NUTS(), 200)
pred = predict(model_Linear(fill(missing, 100), data.x, data.participant, 10), post)
```

It errors for predictions on a grid:

```jl
# Predictions on new data "grid"
grid = DataFrame(x=[0, 50, 100], participant=[missing, missing, missing])
pred2 = predict(model_Linear(fill(missing, 3), grid.x, grid.participant, 3), post)
```
```
ERROR: BoundsError: attempt to access 3-element Vector{Float64} at index [4]
```

I feel like I'm close, but the I can't make sense of the error 🤔

**Reply by torfjelde** [reply]
> As far as I understand, generated_quantities() allows me to access the quantities generated during sampling?

So `generated_quantities` extracts the `return` statement of the `@model` _conditioned_ on the parameter realizations in `chain`. And this is exactly what you want:) 

_However_ one issue with this, is that in your `chain` you'll have variables `μ_intercept_random` that are of length `n` (4, in your case), which might be different for your _prediction_ popluation (3, in your case). This is why you're encountering the `BoundsError`: `predict` tries to find a place to set the variable `μ_intercept_random[4]` (which was present during sampling), but now `μ_intercept_random` is only of length 3 (since you set `n=3`).

So the way to address this is to find a way to make your model _not_ have the participant-level random variables involved when you're predicting (since that will always fail if you're `n` is incorrect). A simple way to do that is

```julia
@model function model_Linear(y, x, participant, n)

    # Priors - Fixed Effects
    μ_intercept ~ Normal(0, 0.5)
    μ_x ~ Normal(0, 0.5)

    σ ~ truncated(Normal(0.0, 1), lower=0)

    # Priors - Random Effects
    μ_intercept_random_sd ~ truncated(Normal(0.0, 0.3), lower=0)
    μ_x_random_sd ~ truncated(Normal(0, 0.3), lower=0)

    # Participant-level priors (we only include these if `participant` is not `nothing`)
    if participant !== nothing
        μ_intercept_random ~ filldist(Normal(0, μ_intercept_random_sd), n)
        μ_x_random ~ filldist(Normal(0, μ_x_random_sd), n)
    end

    for i in 1:length(y)
        # Compute μ
        μ = μ_intercept + μ_x * x[i, 1]
        if participant !== nothing
            μ += μ_intercept_random[participant[i]] + μ_x_random[participant[i]] * x[i, 1]
        end

        # Likelihood
        y[i] ~ Normal(μ, σ)
    end

    # For `generated_quantities`
    return (μ_fixed = μ_intercept .+ μ_x .* x[:, 1],)
end

# Convenience method for when participants aren't available.
model_Linear(y, x) = model_Linear(y, x, nothing, nothing)
```

Then you can do

```julia
# Predictions on new data "grid"
grid = DataFrame(x=[0, 50, 100], participant=[missing, missing, missing])
model_predict = model_Linear(fill(missing, 3), grid.x)
pred2 = predict(model_predict, post)
results = generated_quantities(model_predict, post)
```

**Reply by torfjelde** [reply]
If you really want to make your life a bit nicer, you could also separate the fixed and random effects into two different models:

```julia

"""
    model_linear_base(x)

Model component of linear model with mixed effects.
"""
@model function model_linear_base(x)
    # Priors - Fixed Effects
    μ_intercept ~ Normal(0, 0.5)
    μ_x ~ Normal(0, 0.5)

    # Priors - Random Effects
    μ_intercept_random_sd ~ truncated(Normal(0.0, 0.3), lower=0)
    μ_x_random_sd ~ truncated(Normal(0, 0.3), lower=0)

    return μ_intercept .+ μ_x .* x[:, 1]
end

"""
    model_linear_participant(x, participant, n)

Model component of linear model with participant-specific effects.
"""
@model function model_linear_participant(x, participant, n)
    μ_intercept_random ~ filldist(Normal(0, μ_intercept_random_sd), n)
    μ_x_random ~ filldist(Normal(0, μ_x_random_sd), n)

    return μ_intercept_random[participant] .+ μ_x_random[participant] .* x[:, 1]
end

@model function model_linear(y, x, participant, n)
    @submodel μ_base = model_linear_base(x)
    μ = μ_base

    if participant !== nothing
        @submodel μ_participant = model_linear_participant(x, participant, n)
        μ = μ + μ_participant
    end

    # Observe.
    σ ~ truncated(Normal(0.0, 1), lower=0)
    y .~ Normal.(μ, σ)
end
```

**Reply by DominiqueMakowski** [reply]
Thanks that's nice! I somehow got invested in Turing's development though, and it makes me wonder if the `@submodel` syntax is future-proof (cf. discussions around `returned_quantities` etc)

**Reply by torfjelde** [reply]
> and it makes me wonder if the @submodel syntax is future-proof (cf. discussions around returned_quantities etc)

`@submodel` would not just be removed out of nowhere:) It'll still be around for quite some time, no doubt. But yes, we're thinking about renaming it to `@returned_quantities` or something similar, though the underlying functionality will be identical with _very_ minor syntax changes (specifically, `@submodel x = model` becomes `x = @submodel model`).

So wouldnt' say this should be cause from refraining to use it:)

**Comment by DominiqueMakowski** [comment]
Super sorry to keep annoying with this Tor, but now using the vectorized syntax suggested above it looks like the `predict()` methods got broken...

Here's an MWE:

```jl
using Turing, StatsFuns, DataFrames, Random


@model function mod(y, x1, x2)

    μ_intercept ~ Normal(0, 0.5)
    μ_x1 ~ Normal(1, 0.5)
    μ_x2 ~ Normal(2, 0.5)
    σ ~ truncated(Normal(0.0, 1), lower=0)

    μ = μ_intercept .+ μ_x1 .* x1 .+ μ_x2 .* x2
    y .~ Normal.(μ, σ)
end


fit = mod(rand(10), rand(10), rand(10))
chain = sample(fit, NUTS(), 400)
pred = predict(mod([missing for i in 1:10], rand(10), rand(10)), chain)
```

```
ERROR: MethodError: no method matching loglikelihood(::Normal{Float64}, ::Missing)
The function `loglikelihood` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  loglikelihood(::Distribution{ArrayLikeVariate{N}}, ::AbstractArray{<:Real, M}) where {N, M}
   @ Distributions C:\Users\domma\.julia\packages\Distributions\uuqsE\src\common.jl:444
  loglikelihood(::Distribution{ArrayLikeVariate{N}}, ::AbstractArray{<:AbstractArray{<:Real, N}}) where N
   @ Distributions C:\Users\domma\.julia\packages\Distributions\uuqsE\src\common.jl:461
  loglikelihood(::Model, ::Any)
   @ DynamicPPL C:\Users\domma\.julia\packages\DynamicPPL\aKhfc\src\simple_varinfo.jl:662
  ...

```

My guess is that we should in this example use MvNormal, but is there a way to make it work for likelihood distributions that don't have a multivariate form?

**Reply by torfjelde** [reply]
Ah sorry, that's my bad. `.~` for variables passed as arguments doesn't for like this.

For this, you'll have to use `condition` / `model | (y = ...,)`, i.e.

```julia
@model function mod(x1, x2, ::Type{TV}=Vector{Float64}) where {TV}

    μ_intercept ~ Normal(0, 0.5)
    μ_x1 ~ Normal(1, 0.5)
    μ_x2 ~ Normal(2, 0.5)
    σ ~ truncated(Normal(0.0, 1), lower=0)

    μ = μ_intercept .+ μ_x1 .* x1 .+ μ_x2 .* x2
    y = TV(undef, length(μ))
    y .~ Normal.(μ, σ)
end


fit = mod(rand(10), rand(10)) | (y = rand(10),)
chain = sample(fit, NUTS(), 400)
pred = predict(decondition(fit, :y), chain)
```

This is because we check the left-hand side of `.~` to determine if it's conditioned or not. In the case where you pass `y` as an argument, whether or not it's conditioned is determined by whether or not it equals `missing`. In your case, the left-hand side, i.e. `y`, does _not_ equal `missing` (but is instead a vector of missings), and so is treated as an observation instead of random. 

We could implement it so that it checks each entry of the left-hand side, i.e. `y` in this case, but this will then be equivalent to the for-loop we wrote earlier, in which case the `.~` becomes sort of useless 🤷

**Reply by DominiqueMakowski** [reply]
I see, thanks. 

I feel like this syntax complexifies things quite a bit, unfortunately (I would like to be able to teach these models and code to my students at some point), so I think I'll stick with the loop for now, even though it makes things slower. 

1. Is there a possibility in the future for Turing to detect vector of missing as missing? In other words, is the problem of predictions not working as expected with vectorized syntax a current code limitation that might be fixed in the future or is it related to some fundamental design decisions?
2. Related, is there a possibility that the explicit for-loop syntax (which also has the benefit of not using vectorized operators, which makes it much more readable and writable for non-Julians) reaches similar performance to vectorized syntax? I would naively think that the macro could rewrite the model internally to be pretty much the same (?)
3. What does `, ::Type{TV}=Vector{Float64}) where {TV}` do in the function definition?

**Reply by torfjelde** [reply]
> I feel like this syntax complexifies things quite a bit, unfortunately (I would like to be able to teach these models and code to my students at some point), so I think I'll stick with the loop for now, even though it makes things slower.

I agree with this 👍 It does make things more complicated, and hence I think the for-loop approach generally the preferred way to do things.

> Related, is there a possibility that the explicit for-loop syntax (which also has the benefit of not using vectorized operators, which makes it much more readable and writable for non-Julians) reaches similar performance to vectorized syntax? I would naively think that the macro could rewrite the model internally to be pretty much the same (?)

So the macro in Turing.jl does very little, and this is on purpose:) 
It won't do stuff like rewrite loops, etc. simply because this quickly limits what you can actually put into a loop; we want to support all that is Julia inside a `@model`. This is why we get very nice interoperability with things like DifferentialEquations.jl, etc. for "free".

So overall, I don't think we'd ever start rewriting loops in a model to improve performance.
 
> Is there a possibility in the future for Turing to detect vector of missing as missing? In other words, is the problem of predictions not working as expected with vectorized syntax a current code limitation that might be fixed in the future or is it related to some fundamental design decisions?

This is a much more likely approach we'll take.
This is a design decision we've made for the sake of performance. However, I agree that this is very confusing and in an ideal world, it wouldn't be the case. And so I think either we should fix this somehow (I have a rough idea that might work), _or_ we need to make the limitations of `.~` much clearer. So overall, I think this should be considered a "bug", and we should address it. I'll raise an issue 👍 

> What does , ::Type{TV}=Vector{Float64}) where {TV} do in the function definition?

This is just a performance optimization.
It basically creates an argument to the model which Turing.jl will use alter when suitable.

For example, if you write
```julia
x = Vector(undef, 2)
x .~ Normal()
```
in your model, then `eltype(x)` is `Any` and things will be slow. If you write
```julia
x = Vector{Float64}(undef, 2)
x .~ Normal()
```
this will be fast in most cases, _but_ when we do automatic differentiation you end up using special numbers, e.g. `ForwardDiff.Dual`. Hence, this `Vector{Float64}` won't be able to hold these numbers, causing issues (it used to error, but now we try to expand the container to accomodate the `Dual` type, etc. but this means you get a `Vector{Union{Float64, ...}}` which also results in type instability and performance loss).

So, to address this issue, instead of writing `Vector{Float64}(undef, 2)` in your model, you can add a type argument, as the `::Type{TV}` above and then use this as a replacement in your model like so
```julia
x = TV(undef, 2)
```
Here `TV = Vector{Float64}` in the default case, because that's the default value I used for it above, but when you perform AD, e.g. using ForwardDIff.jl, Turing.jl will automatically convert this `TV` type, which used to be `Vector{Float64}`, into `Vector{Forwarddiff.Dual{...}}`, ensuring that the model is type stable, and so we get great performance.

You can read about this here: https://turinglang.org/docs/tutorials/docs-13-using-turing-performance-tips/#ensure-that-types-in-your-model-can-be-inferred

**Reply by torfjelde** [reply]
But regardless, I would strongly recommend using `condition` and `|` instead of passing observations as arguments. The latter is becoming more and more out of fashion, and the `condition` approach is much better for composability with both the ecosystem and Turing.jl's own functionality.

**Reply by torfjelde** [reply]
And regarding using it to teach others, I'd definitively stick with the simplest stuff like loops over `.~`, etc.

There are tons of performance optimization one can do, and all these "tricks" is just something someone can make use of when they need things to go fast:) Up until that point, those details, e.g. `TV = Vector{Float64}`, are not necessary.

**Reply by DominiqueMakowski** [reply]
Thanks: that one is silly but I have:

```jl
       fit = mod(rand(10), rand(10)) | (y=rand(10),)
```
```
ERROR: invalid method definition in Main: function Base.mod must be explicitly imported to be extended
Stacktrace:
 [1] top-level scope
```
I am `using` Turing, I tried using DynamicPPL as well but I must be missing something 😅

**Reply by torfjelde** [reply]
Ah that one is annoying 😅 But `mod` is a method already defined in Base (it's a base method from the Julia language itself). So when you run this in the REPL, you'll run into this issue as you try to define a _new_ `mod` binding.

So just call your model something else, e.g. `mdl` instead of `mod` or something:)

**Reply by DominiqueMakowski** [reply]
that makes sense, cheers!
